### PR TITLE
Added 2D functions to AStar

### LINKED ISSUE #26128: AStar API using Vector2
Purely QOL suggestion:

Would be nice if the astar class could deal with Vector2 directly. While not a big hassle to convert between Vector2 and Vector3 in GDScript, it would be more fluid to be able to provide Vector2 directly. 

Main use case is pathfinding in 2D.

### PR DESCRIPTION
**Pull request by Chaosus**
QOL change: Added functions which accepts/returns Vector2 instead Vector3

- AStar.add_point_2d
- AStar.get_point_position_2d
- AStar.set_point_position_2d
- AStar.get_closest_point_2d
- AStar.get_closest_position_in_segment_2d
- AStar.get_point_path_2d

Closes #26128

### CODE REVIEW

**Comment by reduz** [comment]
seems fine I guess.. maybe you could make an Astar2D that contains AStar and uses it internally :D

**Comment by Chaosus** [comment]
Yeah, I still not sure whether its a good addition, astar will be in a bit mess after that..

**Comment by menip** [comment]
`AStar2D.add_point()` etc would be more consistent with editor, no?

**Comment by Chaosus** [comment]
Ok, I've created a wrapper class AStar2D, pushed AStar reference to it and makes them friends to access _estimate_cost and _compute_cost methods. I think its ok for now.

File: core/math/a_star.h [hunk 1]
```diff
@@ -142,4 +144,46 @@ class AStar : public Reference {
 	~AStar();
 };
 
+class AStar2D : public Reference {
+	GDCLASS(AStar2D, Reference);
+	AStar *astar;
```
Comment (bojidar-bg) [inline core/math/a_star.h]: I think you don't need a pointer to `astar`, as you have only once instance of it, have it fully defined above, and don't return it from any of the functions.

Comment (Chaosus) [reply core/math/a_star.h]: AStar2D is a wrapper class above AStar so keeping a pointer inside is neccessary to call the AStar functions.

Comment (bojidar-bg) [reply core/math/a_star.h]: I mean, making it a `AStar astar;`, without it being a pointer. Just tested, and it works, though it is probably not the best thing to do.

Comment (Chaosus) [reply core/math/a_star.h]: Ah, ok

**Comment by bojidar-bg** [comment]
`_estimate_cost` is not called by AStar2D, and I guess `_compute_cost` isn't either.
Example:
```gdscript
extends Node
class X extends AStar:
	func _estimate_cost(from_id, to_id):
		prints("AStar", from_id, to_id)
		return get_point_position(from_id).distance_to(get_point_position(to_id))

class X2D extends AStar2D:
	func _estimate_cost(from_id, to_id):
		prints("AStar2D", from_id, to_id)
		return get_point_position(from_id).distance_to(get_point_position(to_id))

func _ready():
	var a = X.new()
	a.add_point(0, Vector3(0, 0, 0))
	a.add_point(1, Vector3(0, 32, 0))
	a.get_point_path(0, 1)
	var b = X2D.new()
	b.add_point(0, Vector2(0, 0))
	b.add_point(1, Vector2(0, 32))
	b.get_point_path(0, 1)
```
Results in:
```
AStar 0 1
```
But should result in:
```
AStar 0 1
AStar2D 0 1
```

**Comment by Chaosus** [comment]
I tried to do it without result - because the AStar2D is a wrapper class and the estimate_cost is already defined in scene_string_names this requires to define another method such as estimate_cost_2d and the _solve method too. It's too clunky I've better remove these methods..

**Comment by anissen** [comment]
How about also adding a dimension-less API to `AStar`? That would be useful for taking advantage of the built-in A* implementation for implementing e.g. [Goal-Oriented Action Programming](http://alumni.media.mit.edu/~jorkin/goap.html) or other algorithms that depend on shortest-path searching in a graph.

**Comment by mhilbrunner** [comment]
@anissen That should be discussed seperately, I think, as it seems outside the scope of the current PR. Maybe open a issue for discussion?

File: doc/classes/AStar2D.xml [hunk 2]
```diff
@@ -0,0 +1,258 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="AStar2D" inherits="Reference" category="Core" version="3.2">
+	<brief_description>
+		AStar class representation that uses 2d-vectors as edges.
+	</brief_description>
+	<description>
+		This is a wrapper for [AStar] class which using 2d-vectors instead 3d.
```
Comment (menip) [inline doc/classes/AStar2D.xml]: uses instead of using

Comment (menip) [reply doc/classes/AStar2D.xml]: `Instead of` not just `instead`.
`This is a wrapper for the [AStar] class which uses 2D vectors instead of 3D vectors.`

Comment (Chaosus) [reply doc/classes/AStar2D.xml]: Fixed

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**